docs/storage/commit_log/commit_log_based_storage_design.md

```diff
@@ -0,0 +1,354 @@
+    batch := kafka.Read("Leaves", leafRange)
+    // ... and store it in our local DB
+    for b := range batch {
+      db.Put("<treeID>/leaves/<b.offset>", b.contents)
+    }
+    UpdateMerkleTreeAndStoreNodes(batch, treeRevision+1)
+    nextSTH.sthOffset = lastOffset
+    StoreSTHInDB(nextSTH)
+    tx.Commit() // flush writes
+    // We may still not be caught up, but that's for the next time around.
+  }
+}
+```
+
+##### Fit with storage interfaces
+
+LogStorage interfaces will need to be tweaked slightly, in particular:
+ - `UpdateSequencedLeaves` should be pulled out of `LeafDequeuer` and moved
+   into a `LeafSequencer` (or something) interface.
+ - It would be nice to introduce a roll-up interface which describes the
+   responsibilities of the "local DB" thing, so that we can compose
+   `commit-queue+local` storage implementations using existing DB impls
+   (or at least not tie this tightly to HBase).
+
+###### TX
+```golang
+
+type splitTX struct {
+   treeID       int64
+   ...
+
+   dbTX         *storage.LogTX // something something handwavy
+   cqTX         *storage.???   // something something handwavy
+
+   dbSTH        *trillian.SignedTreeHead
+   nextSTH      *trillian.SignedTreeHead  // actually something which contains this plus some metadata
+   treeRevision int64
+   sthOffset    int64
+}
+```
+
+###### `Storage.Begin()`
+
+Starts a Trillian transaction, this will do:
+   1. the read of `currentSTH`, `treeRevision`, and `sthOffset` from the DB
+   1. verification of that against it's corresponding entry in Kafka
+
+and return a `LogTX` struct containing these values as unexported fields.
+**The HBase LogTX struct will buffer all writes locally until `Commit` is
+called**, whereupon it'll attempt to action the writes as HBase `PUT` requests
+(presumably it can be smart about batching where appropriate).
+
+```golang
+// Begin starts a Trillian transaction.
+// This will get the latest known STH from the "local" DB, and verify
+// that the corresponding STH in Kafka matches.
+func (ls *CQComboStorage) Begin() (LogTX, error) {
+  // create db and cq "TX" objects
+
+  tx := &splitTX{...}
+
+  // read `dbSTH` (containing `treeRevision` and `sthOffset`) from local DB
+  tx.dbSTH, tx.treeRevision, tx.stdOffset := dbTX.latestSTH()
+
+  // Sanity check that the STH table has what we already know.
+  ourSTH := cqTX.GetSTHAt(tx.sthOffset)
+
+  if ourSTH == nil {
+    return nil, fmt.Errorf("should not happen - local DB has data ahead of STHs topic")
+  }
+  if ourSTH.expectedOffset != dbSTH.sthOffset {
+    return nil, fmt.Errorf("should not happen - local DB committed to invalid STH from topic")
+  }
+  if ourSTH.timestamp != dbSTH.timestamp || ourSTH.tree_size != dbSTH.tree_size {
+    return nil, fmt.Errorf("should not happen - local DB has different data than STHs topic")
+  }
+
+  ...
+
+  return tx, nil
+}
+```
+
+
+###### `DequeueLeaves()`
+Calls to this method ignore `limit` and `cutoff` when there exist newer STHs in
```

Comment:
Maybe we should rework this API then? Also can we assure that the batch size will have a reasonable bound if we ignore the limits?